### PR TITLE
VC: introduce differ to optimize patroller algorithm

### PR DESCRIPTION
**Pull request by zhuangqh**
1. refactor: using differ in configmap checker
2. add differ for patroller in better performance
    - compare tenant and super object in O(N+M)
    - reduce temporary memory copy from lister

Signed-off-by: zhuangqh <zhuangqhc@gmail.com>

### CODE REVIEW

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+func (c ClusterObject) String() string {
+	var ns = c.GetNamespace()
+	if c.Cluster != "" {
+		ns = conversion.ToSuperMasterNamespace(c.Cluster, ns)
+	}
+	return fmt.Sprintf("%s%c%s", ns, '/', c.GetName())
+}
+
+func (c ClusterObject) GetCluster() string {
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:41]: `GetClusterName` is conflict with the buildin one

File: incubator/virtualcluster/pkg/syncer/resources/configmap/checker.go [hunk 2]
```diff
@@ -55,97 +54,74 @@ func (c *controller) PatrollerDo() {
...
+	for _, pCM := range pConfigMaps {
+		pSet.Insert(differ.ClusterObject{Object: pCM})
+	}
 
-	for _, pConfigMap := range pConfigMaps {
-		clusterName, vNamespace := conversion.GetVirtualOwner(pConfigMap)
-		if len(clusterName) == 0 || len(vNamespace) == 0 {
+	vSet := differ.NewDiffSet()
+	for _, cluster := range clusterNames {
+		listObj, err := c.multiClusterConfigMapController.List(cluster)
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/configmap/checker.go:69]: there is a race. if some vc is deleted during this time, the vSet is missing some objects. patrol will requeue them, it's fine

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+	for k := range keySet1 {
+		if keySet2.Has(k) {
+			intersection.Insert(k)
+			keySet1.Delete(k)
+			keySet2.Delete(k)
+		}
+	}
+
+	for k := range keySet1 {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:124]: Can we use WaitGroup to call multiple handler.OnAdd in parallel. The same for handler.OnUpdate and Handler.OnDelete?

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	Len() int
+	Delete(object ClusterObject)
+	Clear()
+	KeySet() sets.String
+	// Difference compute the different keys between caller and callee.
+	// receive a handler and execute to keep consistent with caller.
+	Difference(Differ, Handler)
+}
+
+type container struct {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:64]: How about rename `container` to `diffset`?

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+type Differ interface {
+	Insert(object ...ClusterObject)
+	Has(object ClusterObject) bool
+	Get(key string) ClusterObject
+	Len() int
+	Delete(object ClusterObject)
+	Clear()
+	KeySet() sets.String
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:58]: How about rename `KeySet` to `GetKeys` ?

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+)
+
+type ClusterObject struct {
+	metav1.Object
+	Cluster string
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:30]: How about remove Cluster and add two new fields:

Key string
OwnerCluster string

So the String() logic is moved to the caller site to build the 'key' and the OwnerCluster can be tenant clustername for vObject and super cluster name (e.g., a hard code "super") for pObject.

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ_test.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ_test.go:2]: 2020->2021

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:2]: 2020->2021

File: incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+func (h HandlerFuncs) OnDelete(obj ClusterObject) {
+	if h.DeleteFunc != nil {
+		h.DeleteFunc(obj)
+	}
+}
+
+// FilteringHandler applies the provided filter to all events coming
+// in. If any object match the filter, will skip this func.
+type FilteringHandler struct {
+	FilterFunc func(obj ClusterObject) bool
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go:54]: It seems that we should have two FilterFuncs here: one for object 1 and one for object 2(in OnUpdate() call)

File: incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	Handler    Handler
+}
+
+// OnAdd calls the nested handler only if the filter succeeds
+func (h FilteringHandler) OnAdd(obj ClusterObject) {
+	if !h.FilterFunc(obj) {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go:60]: We should do a nullptr check for the FilterFunc.

**Review by Fei-Guo — COMMENTED** [review]
Overall, I like the idea and appreciate the efforts!

Detailed comments are below.

File: incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go (hunk 9 above)
Comment (zhuangqh) [reply incubator/virtualcluster/pkg/syncer/patrol/differ/handler.go:54]: i would like to keep it simpler. caller can distinguish them in on func.

**Comment by zhuangqh** [comment]
benchmark
```
pkg: sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/patrol/differ
Benchmark_Difference_1000-12    	1000000000	         0.0240 ns/op	       0 B/op	       0 allocs/op
```

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go (hunk 4 above)
Comment (zhuangqh) [reply incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:55]: package name already has 'differ'

File: incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go (hunk 3 above)
Comment (zhuangqh) [reply incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go:141]: good idea, update

**Pull request merged**